<a href="https://colab.research.google.com/github/Panachai-spec007/BSC_DPDM2025/blob/main/midterm_663020285-8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import data

In [18]:
import pandas as pd
import glob
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
sta_hii = pd.read_csv('/content/drive/MyDrive/GreenFN/data_mid_BSC/Stations in the model_HII.csv')
sta_hii.head()

,Node,Lat,Long
0,ACRU,15.788054,104.642235
1,BBHN,9.536913,98.579490
2,BBUA,17.372976,103.984436
3,BCNG,17.386840,103.290870
4,BDAR,14.587073,102.495190


In [7]:
sta_hii.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 385 entries, 0 to 384
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Node    385 non-null    object 
 1   Lat     385 non-null    float64
 2   Long    385 non-null    float64
dtypes: float64(2), object(1)
memory usage: 9.2+ KB


ในไฟล์ Stations in the model_HII.csv จะมีทั้งหมด 3 คอลัมน์

คอลัมน์ NODE คือ ตัวย่อของชื่อสถานี, Lat คือ ละติจูด, Long คือ ลองจิจูด

ดึงไฟล์ HII_daily_rainfall ของทุกเดือนมารวมกัน



In [25]:
base_path = '/content/drive/MyDrive/GreenFN/data_mid_BSC'

# ดึง csv ทั้งหมด
files = glob.glob(base_path + '/**/*.csv', recursive=True)

# รายชื่อไฟล์ที่ไม่ต้องการ
exclude_files = {
    'Stations in the model_HII.csv',
    'Y_monthly_rain_HII_operation.csv',
    'hii_daily_rain.csv'
}

# กรองออก
files = [
    f for f in files
    if os.path.basename(f) not in exclude_files
]

print(f'เหลือไฟล์ที่ใช้จริง {len(files)} ไฟล์')
files[:5]


เหลือไฟล์ที่ใช้จริง 187 ไฟล์


['/content/drive/MyDrive/GreenFN/data_mid_BSC/2025-07/HII_daily_rainfall_2025-07-31.csv',
 '/content/drive/MyDrive/GreenFN/data_mid_BSC/2025-07/HII_daily_rainfall_2025-07-09.csv',
 '/content/drive/MyDrive/GreenFN/data_mid_BSC/2025-07/HII_daily_rainfall_2025-07-10.csv',
 '/content/drive/MyDrive/GreenFN/data_mid_BSC/2025-07/HII_daily_rainfall_2025-07-02.csv',
 '/content/drive/MyDrive/GreenFN/data_mid_BSC/2025-07/HII_daily_rainfall_2025-07-01.csv']

In [33]:
df_list = []

for file in files:
    df = pd.read_csv(file)
    df['source_file'] = file
    df_list.append(df)

all_data = pd.concat(df_list, ignore_index=True)

all_data.head()


,rainfall_datetime,rainfall_datetime_calc,rainfall_value,tele_station_id,tele_station_oldcode,measure_datetime,source_file,station_code,value
0,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,103.0,TB0004,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
1,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,2.2,97.0,CMKK,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
2,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,93.0,HNKA,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
3,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,89.0,CPY005,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
4,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,88.0,HUKT,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN


In [34]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180700 entries, 0 to 180699
Data columns (total 9 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   rainfall_datetime       176886 non-null  object 
 1   rainfall_datetime_calc  176886 non-null  object 
 2   rainfall_value          176882 non-null  float64
 3   tele_station_id         173072 non-null  float64
 4   tele_station_oldcode    168021 non-null  object 
 5   measure_datetime        180700 non-null  object 
 6   source_file             180700 non-null  object 
 7   station_code            3814 non-null    object 
 8   value                   3814 non-null    float64
dtypes: float64(3), object(6)
memory usage: 12.4+ MB


In [49]:
data_2026 = all_data[all_data['year'] == 2026]

if not data_2026.empty:
    print("ข้อมูลสำหรับปี 2026:")
    display(data_2026.head())
    print(f"จำนวนรายการข้อมูลสำหรับปี 2026 ทั้งหมด: {len(data_2026)}")
else:
    print("ไม่พบข้อมูลสำหรับปี 2026 ใน all_data ครับ")

ข้อมูลสำหรับปี 2026:


,rainfall_datetime,rainfall_datetime_calc,rainfall_value,tele_station_id,tele_station_oldcode,measure_datetime,source_file,station_code,value,year,month
88478,2026-01-01 00:00:00+07:00,2025-12-31T07:00:00+07:00,0.0,721.0,DGSL,2025-12-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN,2026,1
88480,2026-01-01 00:00:00+07:00,2025-12-31T07:00:00+07:00,0.0,434.0,CHM002,2025-12-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN,2026,1
88481,2026-01-01 00:00:00+07:00,2025-12-31T07:00:00+07:00,0.0,61.0,LBI002,2025-12-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN,2026,1
88482,2026-01-01 00:00:00+07:00,2025-12-31T07:00:00+07:00,0.0,213.0,BSUM,2025-12-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN,2026,1
88485,2026-01-01 00:00:00+07:00,2025-12-31T07:00:00+07:00,0.6,828.0,TPI007,2025-12-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN,2026,1


จำนวนรายการข้อมูลสำหรับปี 2026 ทั้งหมด: 368


โดยการนำไฟล์ทั้งหมดมารวมแล้วตั้งชื่อเป็น all_data จะมีทั้งหมด 9 คอลัมน์

1. rainfall_datetime วันที่–เวลาของข้อมูลฝน (ต้นฉบับ)

2. rainfall_datetime_calc วันที่–เวลาที่ผ่านการคำนวณ/ปรับแล้ว

3. rainfall_value ค่าปริมาณน้ำฝน

4. tele_station_id รหัสสถานีโทรมาตร (ตัวเลข)

5. tele_station_oldcode รหัสสถานีแบบเก่า / รหัสอ้างอิง

6. measure_datetime วันที่–เวลาที่ทำการวัดจริง

7. source_file ที่มาของไฟล์ข้อมูล

8. station_code รหัสสถานีอีกรูปแบบหนึ่ง

9. value ค่าข้อมูลสำรองหรือค่ามาตรฐานกลาง


# Data Cleaning

## filter Data

In [35]:
print(f"จำนวนสถานีทั้งหมดใน all_data ก่อนกรอง: {all_data['tele_station_oldcode'].nunique()} สถานี")

stations_to_keep = sta_hii['Node'].unique()
all_data_filtered = all_data[all_data['tele_station_oldcode'].isin(stations_to_keep)]

print(f"จำนวนสถานีใน all_data หลังกรอง: {all_data_filtered['tele_station_oldcode'].nunique()} สถานี")

all_data = all_data_filtered.copy()
all_data.head()

จำนวนสถานีทั้งหมดใน all_data ก่อนกรอง: 1514 สถานี
จำนวนสถานีใน all_data หลังกรอง: 384 สถานี


,rainfall_datetime,rainfall_datetime_calc,rainfall_value,tele_station_id,tele_station_oldcode,measure_datetime,source_file,station_code,value
2,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,93.0,HNKA,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
7,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,109.0,SWNG,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
8,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,110.0,PLTL,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
11,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,87.0,WTSG,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN
12,2025-08-01 00:00:00+07:00,2025-08-01T07:00:00+07:00,0.0,115.0,RAY001,2025-07-31,/content/drive/MyDrive/GreenFN/data_mid_BSC/20...,NaN,NaN


จำนวนสถานีทั้งหมดใน all_data ก่อนกรอง: 1514 สถานี

จำนวนสถานีใน all_data หลังกรอง: 384 สถานี

นำข้อมูลที่ filter แล้วไปใช้

## Check Missing

In [37]:
all_data.isnull().sum()

,0
rainfall_datetime,0
rainfall_datetime_calc,0
rainfall_value,1
tele_station_id,1023
tele_station_oldcode,0
measure_datetime,0
source_file,0
station_code,66275
value,66275


In [36]:
missing_data_by_station = all_data.groupby('tele_station_oldcode')['rainfall_value'].apply(lambda x: x.isnull().sum())
total_data_by_station = all_data.groupby('tele_station_oldcode')['rainfall_value'].count()

missing_df = pd.DataFrame({
    'missing_count': missing_data_by_station,
    'total_count': total_data_by_station + missing_data_by_station # Include missing in total count
})
missing_df['missing_proportion'] = missing_df['missing_count'] / missing_df['total_count']

# Sort by missing proportion in descending order for easier review
missing_df = missing_df.sort_values(by='missing_proportion', ascending=False)

print("ข้อมูลสูญหาย (Missing Data) ในแต่ละสถานี:")
print(missing_df.head(10)) # Display top 10 stations with most missing data

print("\nสรุปข้อมูลสูญหายโดยรวม:")
total_missing = all_data['rainfall_value'].isnull().sum()
total_entries = len(all_data)
total_missing_proportion = total_missing / total_entries
print(f"จำนวนข้อมูลสูญหายทั้งหมด: {total_missing} จาก {total_entries} แถว")
print(f"สัดส่วนข้อมูลสูญหายทั้งหมด: {total_missing_proportion:.2%}")

ข้อมูลสูญหาย (Missing Data) ในแต่ละสถานี:
                      missing_count  total_count  missing_proportion
tele_station_oldcode                                                
KKTO                              1          147            0.006803
NKNN                              0          184            0.000000
SITP                              0          183            0.000000
SISU                              0          184            0.000000
SGKW                              0          136            0.000000
SENM                              0          184            0.000000
SBT1                              0          184            0.000000
SAY002                            0          184            0.000000
SAY001                            0          176            0.000000
RNGH                              0          184            0.000000

สรุปข้อมูลสูญหายโดยรวม:
จำนวนข้อมูลสูญหายทั้งหมด: 1 จาก 66275 แถว
สัดส่วนข้อมูลสูญหายทั้งหมด: 0.00%


### คัดสถานีที่มีข้อมูลสูญหาย >= 20% ออก

In [38]:
# Identify stations to drop (missing proportion >= 20%)
stations_to_drop = missing_df[missing_df['missing_proportion'] >= 0.20].index.tolist()

# Filter all_data to keep only stations with missing proportion < 20%
all_data_cleaned = all_data[~all_data['tele_station_oldcode'].isin(stations_to_drop)].copy()

# Get remaining stations
remaining_stations = all_data_cleaned['tele_station_oldcode'].unique()

print(f"สถานีที่ถูกคัดออกเนื่องจากมีข้อมูลสูญหาย >= 20% ({len(stations_to_drop)} สถานี):\n{stations_to_drop}\n")
print(f"สถานีที่เหลืออยู่ ({len(remaining_stations)} สถานี):\n{remaining_stations[:10]}... (แสดงเพียง 10 สถานีแรก)\n")

print(f"จำนวนแถวใน all_data ก่อนคัดออก: {len(all_data)}")
print(f"จำนวนแถวใน all_data หลังคัดออก: {len(all_data_cleaned)}")

all_data = all_data_cleaned.copy()

สถานีที่ถูกคัดออกเนื่องจากมีข้อมูลสูญหาย >= 20% (0 สถานี):
[]

สถานีที่เหลืออยู่ (384 สถานี):
['HNKA' 'SWNG' 'PLTL' 'WTSG' 'RAY001' 'KHHM' 'BNKN' 'LNRG' 'CHPA' 'VLGE27']... (แสดงเพียง 10 สถานีแรก)

จำนวนแถวใน all_data ก่อนคัดออก: 66275
จำนวนแถวใน all_data หลังคัดออก: 66275


### เติม missing ด้วยค่ากลาง

In [41]:
median_rainfall = all_data['rainfall_value'].median()
all_data['rainfall_value'] = all_data['rainfall_value'].fillna(median_rainfall)

print(f"เติมข้อมูลสูญหายใน 'rainfall_value' ด้วยค่ามัธยฐาน: {median_rainfall:.2f}")
print("ตรวจสอบข้อมูลสูญหายอีกครั้ง:")
print(all_data.isnull().sum())

เติมข้อมูลสูญหายใน 'rainfall_value' ด้วยค่ามัธยฐาน: 0.00
ตรวจสอบข้อมูลสูญหายอีกครั้ง:
rainfall_datetime             0
rainfall_datetime_calc        0
rainfall_value                0
tele_station_id            1023
tele_station_oldcode          0
measure_datetime              0
source_file                   0
station_code              66275
value                     66275
dtype: int64


**อภิปราย : ทำไมถึงใช้ median แทน mean**

   การเลือกใช้ค่ามัธยฐาน (median) แทนค่าเฉลี่ย (mean) ในกรณีของข้อมูลปริมาณน้ำฝนนั้น มีเหตุผลหลักมาจากลักษณะการกระจายตัวของข้อมูลปริมาณน้ำฝน
   
  1. ข้อมูลปริมาณน้ำฝนมักมีการกระจายตัวแบบไม่สมมาตร (Skewed Distribution): โดยทั่วไปแล้ว ในแต่ละวันจะมีวันที่ไม่มีฝนตกเลย (ค่า 0) หรือมีฝนตกน้อยเป็นส่วนใหญ่ และมีวันที่ฝนตกหนักมากเป็นส่วนน้อย ทำให้ข้อมูลมีการกระจายตัวไปทางขวา (right-skewed)
   
  2. ความไวต่อค่าผิดปกติ (Outliers): ค่าเฉลี่ย (mean) จะอ่อนไหวต่อค่าผิดปกติหรือค่าที่สูงมากๆ ได้ง่าย หากมีวันที่ฝนตกหนักผิดปกติเพียงไม่กี่วัน ค่าเฉลี่ยจะถูกดึงให้สูงขึ้น ซึ่งอาจทำให้ค่าเฉลี่ยไม่สะท้อนปริมาณน้ำฝน 'โดยทั่วไป' ได้ดีนัก
   
  3. ความทนทานของมัธยฐาน (Robustness of Median): ค่ามัธยฐาน (median) คือค่ากลางของข้อมูลที่เรียงลำดับแล้ว ดังนั้นค่าที่สูงหรือต่ำมากๆ จะไม่ส่งผลกระทบต่อค่ามัธยฐานมากเท่าค่าเฉลี่ย การใช้มัธยฐานจึงช่วยให้ค่าที่นำมาเติมมีความสมจริงและเป็นตัวแทนของปริมาณน้ำฝนส่วนใหญ่ได้ดีกว่า โดยเฉพาะอย่างยิ่งเมื่อมัธยฐานเป็น 0.00 ซึ่งหมายความว่ากว่าครึ่งหนึ่งของข้อมูลปริมาณน้ำฝนคือ 0 มม. หรือน้อยกว่านั้น

In [43]:
Y_monthly_rain_HII_operation = pd.read_csv('/content/drive/MyDrive/GreenFN/data_mid_BSC/Y_monthly_rain_HII_operation.csv')

# คำนวณผลรวมปริมาณน้ำฝนรายเดือน

In [51]:
# Convert 'rainfall_datetime_calc' to datetime objects
#ต้องใช้ 'rainfall_datetime_calc' เพราะจะได้คำนวณวันเวลาได้ถูกต้อง
all_data['rainfall_datetime_calc'] = pd.to_datetime(all_data['rainfall_datetime_calc'])

# Extract year and month
all_data['year'] = all_data['rainfall_datetime_calc'].dt.year
all_data['month'] = all_data['rainfall_datetime_calc'].dt.month

# Group by station, year, and month, then sum 'rainfall_value'
monthly_rain = all_data.groupby(['tele_station_oldcode', 'year', 'month'])['rainfall_value'].sum().reset_index()

# Rename columns to match Y_monthly_rain_HII_operation
monthly_rain = monthly_rain.rename(columns={
    'tele_station_oldcode': 'Node',
    'rainfall_value': 'rain'
})

# Reorder columns to match Y_monthly_rain_HII_operation
monthly_rain = monthly_rain[['Node', 'rain', 'month', 'year']]

print("Monthly rainfall data formatted as requested:")
print(monthly_rain.head())
print("\nInfo of the new DataFrame:")
monthly_rain.info()

Monthly rainfall data formatted as requested:
   Node   rain  month  year
0  ACRU   83.0      7  2025
1  ACRU  164.6      8  2025
2  ACRU  248.8      9  2025
3  ACRU   48.4     10  2025
4  ACRU   52.2     11  2025

Info of the new DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2213 entries, 0 to 2212
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Node    2213 non-null   object 
 1   rain    2213 non-null   float64
 2   month   2213 non-null   int32  
 3   year    2213 non-null   int32  
dtypes: float64(1), int32(2), object(1)
memory usage: 52.0+ KB


In [54]:
monthly_rain.head(7)

,Node,rain,month,year
0,ACRU,83.0,7,2025
1,ACRU,164.6,8,2025
2,ACRU,248.8,9,2025
3,ACRU,48.4,10,2025
4,ACRU,52.2,11,2025
5,ACRU,0.2,12,2025
6,BBHN,390.6,7,2025


In [56]:
# ระบุพาธสำหรับบันทึกไฟล์ CSV
output_path = '/content/drive/MyDrive/GreenFN/ส่งคำตอบ midterm/monthly_rainfall_summary.csv'

# บันทึก DataFrame เป็นไฟล์ CSV
monthly_rain.to_csv(output_path, index=False)

print(f"บันทึกไฟล์ 'monthly_rainfall_summary.csv' ที่ {output_path} เรียบร้อยแล้วครับ")

บันทึกไฟล์ 'monthly_rainfall_summary.csv' ที่ /content/drive/MyDrive/GreenFN/ส่งคำตอบ midterm/monthly_rainfall_summary.csv เรียบร้อยแล้วครับ


#ลิงค์ไฟล์คำตอบ

ไฟล์ผลลัพธ์ https://drive.google.com/file/d/12vaU-PqkEePkN9sI1c0ZgXSlemGBLhBf/view?usp=sharing

ไฟล์ prompt https://drive.google.com/file/d/1H1bmaHvztpL7d9B3yP10ZdYdv33TUCuG/view?usp=sharing